### ISSUE TITLE
ipaddress IPv4Network and IPv6Network need specialized reverse_pointer property

### ISSUE THREAD

**Issue by pemensik** [python/cpython#125641]
# Bug report

### Bug description:

```python
# Add a code block here, if required
from ipaddress import ip_network

# Generates now:
# '8.4./.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.8.b.d.0.1.0.0.2.ip6.arpa'
assert(ip_network("2001:db8::/48").reverse_pointer == ['0.0.0.0.8.b.d.0.1.0.0.2.ip6.arpa'])

# Generates now:
# '0/29.2.0.192.in-addr.arpa'
assert(ip_network("192.0.2.0/29").reverse_pointer == [
'0.2.0.192.in-addr.arpa', '1.2.0.192.in-addr.arpa', '2.2.0.192.in-addr.arpa', '3.2.0.192.in-addr.arpa', '4.2.0.192.in-addr.arpa',
'5.2.0.192.in-addr.arpa', '6.2.0.192.in-addr.arpa', '7.2.0.192.in-addr.arpa'])

```

IPv4Network and IPv6Network get their implementation from _BaseAddress. But for network, it does not work as it should. It returns broken nonsense instead of helping data. Problem is unlike normal IP address, for a network range, it may return just single string for prefix lengths divisible by 8 for IPv4 and 4 for IPv6. But for other prefixes, it should return a list of domain names used.

I have implemented working generator for a list result at https://github.com/InfrastructureServices/dnsconfd/pull/70
I think something similar should be used in base ipaddress directly. If reverse_pointer should not be fixed, it should be removed from networks instead.

Related to #123409, but that is not exactly about networks. Can be verified a bit using ``ipcalc --reverse-dns``, but even that crashes on undivisible ipv6 prefixes.

### CPython versions tested on:

3.13

### Operating systems tested on:

Linux

**Comment by Zheaoli** [comment]
Seems, we need to get more detail about the reverse pointer from the RFC. 

@picnixz plz assign this issue to me, I will take handle of it .

**Comment by picnixz** [comment]
You can take care of the issue, I'm not here for the next days. I didn't know that the reverse pointer for network was different (clearly not my field of research) so I'm happy if you can fix the bug!

**Comment by pemensik** [comment]
There is no such thing as a reverse pointer for networks in dns itself. But it is related to how those domains can be delegated. Meaning where RFC 1035 can cut between zones operated by someone else. [rfc2317](https://datatracker.ietf.org/doc/html/rfc2317) specifies recommended classless delegation using DNAME or CNAMEs in DNS.

[Classless in-addr.arpa delegation](https://datatracker.ietf.org/doc/html/rfc2317#section-4) describes how to delegate ``1.0/25.2.0.192``, therefore ``192.0.2.0/25`` network. Example from that page.

```
   $ORIGIN 2.0.192.in-addr.arpa.
   @       IN      SOA     my-ns.my.domain. hostmaster.my.domain. (...)
   ;...
   ;  <<0-127>> /25
   0/25            NS      ns.A.domain.
   0/25            NS      some.other.name.server.
   ;
   1               CNAME   1.0/25.2.0.192.in-addr.arpa.
   2               CNAME   2.0/25.2.0.192.in-addr.arpa.
   3               CNAME   3.0/25.2.0.192.in-addr.arpa.
```

This means that in  ``2.0.192.in-addr.arpa.`` zone, one needs to send ``['0.2.0.192.in-addr.arpa.', '1.2.0.192.in-addr.arpa.', ... '127.2.0.192.in-addr.arpa.']`` domains to some other domain. Those are the same, what I want generated by my proposal of . Special name ``0/25.2.0.192.in-addr.arpa.`` is done to be able to delegate part of ``2.0.192.in-addr.arpa.`` domain to DNS server operated by someone else. In that names `` 1.0/25.2.0.192.in-addr.arpa.`` to `` 127.0/25.2.0.192.in-addr.arpa.`` can be administered by different entity than original zone ``2.0.192.in-addr.arpa.``.

There is no similar proposal for IPv6, because common IPv6 prefixes assigned to people are divided by 4 with zero remainder. That is /48, /52, /56, /60, /64. But prefix like /54 is still valid, but needs multiple domains to hold reverse addresses contained in that prefix.

**Comment by ZeroIntensity** [comment]
This is affecting all bugfix+ versions (3.12+), right?

**Comment by Zheaoli** [comment]
I think we just need to backport to 3.13 if this is necessary.

**Comment by Zheaoli** [comment]
I might want to ask some questions about this issue

in your test code

```python
# '8.4./.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.8.b.d.0.1.0.0.2.ip6.arpa'
assert(ip_network("2001:db8::/48").reverse_pointer == ['0.0.0.0.8.b.d.0.1.0.0.2.ip6.arpa'])
```

I'm not sure the `0.0.0.0.8.b.d.0.1.0.0.2.ip6.arpa` behavior is defined in the RFC. So I would like to hear about your thought about IPV6 part.

**Comment by pemensik** [comment]
When I use ``ipcalc 2001:db8::/48``, I get these ranges.

```
HostMin:	2001:db8::
HostMax:	2001:db8:0:ffff:ffff:ffff:ffff:ffff
```

So yes, it seems in my code generating this was wrong. ``'0.8.b.d.0.1.0.0.2.ip6.arpa'`` should be correct prefix common for all those reversed domains contained in that range. ``network_address`` and ``broadcast_address`` properties of _BaseNetwork should contain similar values.

**Comment by pemensik** [comment]
In case prefix is not divisible by 4, subnets contained in parent divisible by 4 need to be generated. In fact network_address.reverse_pointer can be reused and just proper number of label splitted from it. Our example is at:
 
https://github.com/InfrastructureServices/dnsconfd/blob/761c5fd37bd0d0b163a21fb6cabed14142004fcf/dnsconfd/network_objects/server_description.py#L190

**Comment by Forst** [comment]
There's already a couple of issues with associated pull requests open for the same problem:

* #74713
* #91601

The first issue in particular has some more ideas and discussion on the matter, where there a few other options on how to treat `reverse_pointer` for network objects:

* Generate a single, more general reverse pointer (e.g. `192.0.2.128/25` -> `2.0.192.in-addr.arpa.`)
* Raise an exception when generating reverse pointers for networks that don't fall on the DNS record granularity boundary (8 bits for IPv4, 4 bits for IPv6)
* Remove `reverse_pointer` for network objects altogether

The suggestion in this issue to return a list of one or more pointers is interesting, but a bit tricky in my opinion. For one, it may result in confusion where the address objects return a single string, but networks return a list for the same attribute. Also, the name `reverse_pointer` suggests it will return one pointer of some sort, rather than a list.

RFC 2317 mentioned above only covers prefixes larger than `/24` in IPv4. It covers neither smaller prefixes nor IPv6, and that's something we should also think about.

**Comment by picnixz** [comment]
@Zheaoli are you still working on this?

**Comment by pemensik** [comment]
> The first issue in particular has some more ideas and discussion on the matter, including a few other options on how to treat `reverse_pointer` for network objects:
> 
>     * Generate a single, more general reverse pointer (e.g. `192.0.2.128/25` -> `2.0.192.in-addr.arpa.`)

Nope, that would be wrong. There might be a helper to this, but the given reverse pointer is for 192.0.2.0/24 network, not for the entry specified. Returning partially invalid data is not a good idea, especially if that desire were not explicitly requested. It may work for something until it creates visible errors. Wrong way IMO.

>     * Raise an exception when generating reverse pointers for networks that don't fall on the DNS record granularity boundary (8 bits for IPv4, 4 bits for IPv6)

And when one does have networks with non-clean divisible block, he should not be able to work with them, just because we may need change return type? Why? Is there anything wrong with prefixes not divisible by 8 or 4?
> 
>     * Remove `reverse_pointer` for network objects altogether

Well, it it would mean we rename it to ``reverse_pointers`` instead, okay. After all, network is a group of addresses and that it needs a group of names for them, it should not be surprising. But just removing without any replacement, that is a bad idea.
> 
> 
> The suggestion in this issue to return a list of one or more pointers is interesting, but a bit tricky in my opinion. For one, it may result in confusion where the address objects return a single string, but networks return a list for the same attribute. 

Network is a group of address. Unlike single address it represents multiple of addresses. Well then multiple of reverse pointers should be documented, but not considered surprising.

> Also, the name `reverse_pointer` suggests it will return one pointer of some sort, rather than a list.

I would be okay if reverse_pointer were kept and throwing Error if not divisible, but have reverse_pointers property providing always working list or set for any prefix.

> 
> RFC 2317 mentioned above only covers prefixes larger than `/24` in IPv4. It covers neither smaller prefixes nor IPv6, and that's something we should also think about.

Yes, RFC 2317 is just a hint how to configure DNS server to delegate parts of IPv4 space to other servers. It addresses just too big chunks of 8 bits per IPv4 number, which lacks better granularity. It relies on code embedded in DNS servers to generate a list from network 192.0.2.128/26 on the server side. Not always that is enough and we want ability to generate that list ourselves in any case.

RFC 2317 did not have to deal with prefix smaller than 24, because those were possible to be gelegated just normal DNS delegation with common labels. But prefixes from 24 to 32, there are only final records, which should not be a CNAME.

**Comment by Forst** [comment]
> I would be okay if reverse_pointer were kept and throwing Error if not divisible, but have reverse_pointers property providing always working list or set for any prefix.

This sounds good to me! One could perhaps, for the sake of consistency, even implement `reverse_pointers` at the base class level to make it work for IP addresses as well, returning a list of a single item?
